### ISSUE TITLE
False-positive detection of assignment/condition

### ISSUE THREAD

**Issue by tfrommen** [WordPress-Coding-Standards/WordPress-Coding-Standards#1227]
**Error Code:** `WordPress.CodeAnalysis.AssignmentInCondition.Found`

PHP 7, and WPCS as well as PHPCS latest stable.

----

Hey there, me again! 👋 🙂 

I just stumbled over some _interesting_ issue.

Consider the following code, to be found in `foo.php`:

```php
<?php
/**
 * Foo.
 *
 * @package foo
 */

( function () {
	$foo = 42;

	return 1 === 2 ? 'a' : 'b';
} );
```

The code is the best I could come up in trying to simplify my real code example, which is a lot more complex). Actually, the callback is being fed to some constructor, but just wrapping it with parentheses already does the trick (or bug).

When running `$ phpcs foo.php --standard=WordPress`, we get the following output:

```text
FILE: /path/to/foo.php
----------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------------------------------------
 9 | WARNING | Variable assignment found within a condition. Did you mean to do a comparison?
----------------------------------------------------------------------------------------------------

Time: 526ms; Memory: 8Mb
```

L9 is `$foo = 42;`, so where is the condition?

I played around with this, and all of the following _variations_ do not trigger the sniff:

```php
// No ternary.
( function () {
	$foo = 42;

	return 1 === 2;
} );

// No variable definition.
( function () {

	return 1 === 2 ? 'a' : 'b';
} );

// Store closure in a variable.
$foo = function () {
	$foo = 42;

	return 1 === 2 ? 'a' : 'b';
};

/** Don't put function in parentheses (which means I have to give it a name, of course). */
function foo() {
	$foo = 42;

	return 1 === 2 ? 'a' : 'b';
}

// Use the function body (as a block).
{
	$foo = 42;

	return 1 === 2 ? 'a' : 'b';
}
```

I am going to ignore the Sniff/message for now, but it would be great to have this fixed. :)

Thanks,
Thorsten

**Comment by dingo-d** [comment]
I'm just curious because I've never seen this in php (I've seen similar thing in JS)

```php
( function () {
    $foo = 42;
    
    return 1 === 2 ? 'a' : 'b';
} );
```

What do the braces around the function mean? Some kind of a closure?

**Comment by tfrommen** [comment]
@dingo-d I tried to condense the snippet as good as possible, but maybe that was not so intuitive.

In this case, the parentheses just _group_ the code, nothing else. But you could easily make it a function call, for example, like so:

```php
<?php
/**
 * Foo.
 *
 * @package foo
 */

call_user_func( function () {
	$foo = 42;

	return 1 === 2 ? 'a' : 'b';
} );
```

This results in the same PHPCS/WPCS warning. Maybe that's easier to understand.

**Comment by dingo-d** [comment]
Oh, ok, I thought that was some new php feature I don't know about :D

**Comment by jrfnl** [comment]
Hi @tfrommen Thanks for reporting this. The "assignment in condition in ternary" part of the sniff is the part for which I did expect some issues, so thanks for giving me another test case :sunglasses: 

I will look into it when I have some time, but it may be a while before I do.

**Comment by paulschreiber** [comment]
I'm seeing this problem with WPCS 0.14 and PHPCS 3.1.1 with this code:
```
// Replace any viz tables with iframe representations.
$content = preg_replace_callback(
	'/(?s)<table class="viz full.+?<\/table>/',
	function( $matches ) {

		$table = $matches[0];

		// For smaller tables, use a shorter iframe.
		$rowcount = substr_count( $table, '<tr>' );
		$height   = $rowcount < 6 ? '100' : '250';
		return '
			<amp-iframe
			height="' . $height . '"
			class="table-viz-iframe"
			frameborder="1"
			sandbox="allow-scripts"
			srcdoc="<html><head><style>...</style></head><body>' . $table ) . '</body></html>">
			</amp-iframe>
		';
	},
	$content
);
```

**Comment by jrfnl** [comment]
@paulschreiber @tfrommen Thank you both for giving me example code to work with. I believe this should be fixed with PR #1249

**Comment by jrfnl** [comment]
FYI: I've just pushed an improved fix + additional unit tests. The ternary check in the sniff now also has its own errorcode.

**Comment by tfrommen** [comment]
Thanks, @jrfnl - I actually confirmed this is working as (I) needed, but forgot to report back here.

**Comment by jrfnl** [comment]
@tfrommen Thanks for letting me know. Glad it works.
